File: ts-src/src/algo/bloom.ts
import { xxh64Raw } from './xxhash.js';

const MAX_BITS = 2 ** 31 - 1;
const encoder = new TextEncoder();

function optimalM(n: number, p: number): number {
  return Math.min(MAX_BITS, Math.max(1, Math.floor(-n * Math.log(p) / (Math.log(2) ** 2))));
}

function optimalK(m: number, n: number): number {
  return Math.max(1, Math.floor(m * Math.log(2) / n));
}

function hash64(data: string, seed: bigint): bigint {
  return xxh64Raw(encoder.encode(data), seed);
}

/**
 * Space-efficient probabilistic set for membership testing.
 *
 * Uses Kirsch–Mitzenmacher double-hash technique.
 * False positives possible; false negatives are not.
 */
export class BloomFilter {
  private m: number;
  private k: number;
  private bits: Uint8Array;

  constructor(expectedItems: number, fpRate = 0.01) {
    if (expectedItems < 1) throw new Error(`expectedItems must be >= 1, got ${expectedItems}`);
    if (fpRate <= 0 || fpRate >= 1) throw new Error(`fpRate must be in (0, 1), got ${fpRate}`);
    const n = expectedItems;
    this.m = optimalM(n, fpRate);
    this.k = Math.min(optimalK(this.m, n), this.m);
    this.bits = new Uint8Array(Math.ceil(this.m / 8));
  }

  private hashes(item: string): number[] {
    const h1 = hash64(item, 0n);
    const h2 = hash64(item, 0x9e3779b97f4a7c15n);
    const result: number[] = [];
    const bigM = BigInt(this.m);
    for (let i = 0; i < this.k; i++) {
      result.push(Number((h1 + BigInt(i) * h2) % bigM));
    }
    return result;
  }

  add(item: string): void {
    for (const pos of this.hashes(item)) {
      this.bits[Math.floor(pos / 8)]! |= 1 << (pos % 8);
    }
  }

  mightContain(item: string): boolean {
    for (const pos of this.hashes(item)) {
      if (!(this.bits[Math.floor(pos / 8)]! & (1 << (pos % 8)))) return false;
    }
    return true;
  }

  serialize(): Buffer {
    const header = Buffer.alloc(8);
    header.writeUInt32LE(this.m, 0);
    header.writeUInt32LE(this.k, 4);
    return Buffer.concat([header, Buffer.from(this.bits)]);
  }

  /** @internal Construct with pre-computed params (used by deserialize). */
  private static fromRaw(m: number, k: number, bits: Uint8Array): BloomFilter {
    const bf = new BloomFilter(1);
    bf.m = m;
    bf.k = k;
    bf.bits = bits;
    return bf;
  }

  static deserialize(data: Buffer): BloomFilter {
    if (data.length < 8) throw new Error(`Bloom filter data too short: ${data.length} bytes`);
    const m = data.readUInt32LE(0);
    const k = data.readUInt32LE(4);
    if (m < 1 || k < 1) throw new Error(`Invalid Bloom filter params: m=${m}, k=${k}`);
    const expectedBytes = Math.ceil(m / 8);
    if (data.length - 8 !== expectedBytes) {
      throw new Error(`Bloom filter size mismatch: got ${data.length - 8}, expected ${expectedBytes}`);
    }
    return BloomFilter.fromRaw(m, k, new Uint8Array(data.subarray(8)));
  }
}

File: src/sync/bloom.py
"""
Bloom filter for probabilistic set membership.

Used in youdao note sync for efficient "seen" tracking (e.g. deduplication).
False positives possible; false negatives are not.
"""

import math
import struct
from typing import List

import xxhash


def _optimal_m(n: int, p: float) -> int:
    """m = -n * ln(p) / (ln(2)^2)"""
    return max(1, int(-n * math.log(p) / (math.log(2) ** 2)))


def _optimal_k(m: int, n: int) -> int:
    """k = m/n * ln(2)"""
    return max(1, int(m * math.log(2) / n))


class BloomFilter:
    """
    Space-efficient probabilistic set for membership testing.

    Uses xxhash with different seeds for k independent hash functions.
    Serialization format: 8-byte header (m, k as uint32) + raw bit array.
    """

    def __init__(self, expected_items: int, fp_rate: float = 0.01) -> None:
        if not (0 < fp_rate < 1):
            raise ValueError(f"fp_rate must be in (0, 1), got {fp_rate}")
        n = max(1, expected_items)
        self._m = _optimal_m(n, fp_rate)
        self._k = min(_optimal_k(self._m, n), self._m)
        self._bits = bytearray((self._m + 7) // 8)

    def _hashes(self, item: str) -> List[int]:
        """Kirsch-Mitzenmacher: h_i = h1 + i*h2 mod m for better independence."""
        data = item.encode("utf-8")
        h1 = xxhash.xxh64(data, seed=0).intdigest()
        h2 = xxhash.xxh64(data, seed=0x9E3779B97F4A7C15).intdigest()
        return [(h1 + i * h2) % self._m for i in range(self._k)]

    def add(self, item: str) -> None:
        """Add item to the filter."""
        for pos in self._hashes(item):
            self._bits[pos // 8] |= 1 << (pos % 8)

    def might_contain(self, item: str) -> bool:
        """Return True if item may be in the set (may have false positives)."""
        for pos in self._hashes(item):
            if not (self._bits[pos // 8] & (1 << (pos % 8))):
                return False
        return True

    def serialize(self) -> bytes:
        """Serialize to bytes for storage or transmission."""
        return struct.pack("<II", self._m, self._k) + bytes(self._bits)

    @classmethod
    def deserialize(cls, data: bytes) -> "BloomFilter":
        """Reconstruct from serialized bytes."""
        if len(data) < 8:
            raise ValueError(f"Bloom filter data too short: {len(data)} bytes (need ≥8)")
        m, k = struct.unpack("<II", data[:8])
        if m < 1 or k < 1:
            raise ValueError(f"Invalid Bloom filter parameters: m={m}, k={k} (both must be ≥1)")
        expected_bytes = (m + 7) // 8
        actual_bytes = len(data) - 8
        if actual_bytes != expected_bytes:
            raise ValueError(
                f"Bloom filter bit array size mismatch: got {actual_bytes}, expected {expected_bytes}"
            )
        bf = object.__new__(cls)
        bf._m = m
        bf._k = k
        bf._bits = bytearray(data[8:])
        return bf
